### DISCUSSION TITLE
[Message-Convertor] OSI

### DISCUSSION THREAD

**Discussion by JonasHablitzelAVLD** [foxglove/community#867] [Q&A] [answered]
Hello, 

i am currently working on a POC of using mcap and Foxglove with a message extension to visualize [OSI data](https://github.com/OpenSimulationInterface/open-simulation-interface). OSI is a Protobuf based  generic interface standard for the exchange of data between simulation tools like [Esmini](https://github.com/esmini/esmini). 
The Protobuf SensorData contains multiple elemts like objects(Bboxes) or raw Sensor output (Pointcloud) in one Protobuf message!
As the message converter is mainly designed to have one topic in and one topic out this makes the implementation of the OSI standard quiet hard, as one OSI message has N internal signals. 
Is therefor a possibility to translate one topic into multiple topics for the 3d panel through the message extension? Currently i put everything in a SceneUpdate, which does not distinguishe between the Pointclouds or the Objects properly.

**Comment by defunctzombie** [comment, marked as answer]
You have a few options depending on what workflows you are looking to accomplish.

Message converters don't deal with topics but focus on schemas. They convert from one schema to another schema. You can register multiple converters for the same input schema to different output schemas. With message converters, panels like 3d, image, and map still show your original topic name in topic lists.

User-scripts are another way to convert messages. But unlike converters, which deal with schemas on any topic, user-scripts you specify input and output topics. So if the workflow you want is to create new "topics" from the single protobuf input topic you would use a user-script.

**Reply by JonasHablitzelAVLD** [reply]
Hi thanks for the fast Reply!

from a general point the message converter would fit better for me as i want to map a schema to a new schema and not to use a topic name.

Maybe as an example from the ros world the message would look similar to this:
```
std_msgs/Header header  
sensor_msgs/Image[] images 
sensor_msgs/PointCloud2[] pointclouds
```

With the current capabilities of Foxglove i think the best solution will be to use a user script which publishes the individual array elements plus a message converter for the new produced schemas. What is your opinion on this?

Some additional questions:
Is there a plan to support a message conveter which is able to tinker with the topic names and has a 1..N schema mapping?
Is there a plan to import own packages into the user script?
Is there a plan to support also schema names as selectors and not only message names on user scripts?

**Reply by defunctzombie** [reply]
In the above situation where you have an input message of the form:

```
std_msgs/Header header  
sensor_msgs/Image[] images 
sensor_msgs/PointCloud2[] pointclouds
```

This is not ideally suitable for message converters because there's no schema we have that can output multiple images in a single message or multiple point clouds. I now understand better what you mean by the `N` in `1..N` schema mapping. For this situation a user-script is the appropriate tool today. You will be able to create new topics for each of the images and point cloud entries in the array. This is likely not perfect for you if the number of images or pointclouds is not known ahead of time. So for today I recommend a user script that will produce an output topic for each index of your array. So you would have a script that accepts the input message above and outputs a topic like `/output/image_0` and that script would grab the 0'th image from the input message and output it on that topic. Not the most elegant set of scripts to write but it will work.

We do have plans to support message converters that can tinker with topic names - like user-scripts do. Probably by providing a way to offer what user-scripts do as extensions like we do for message converters. Once this is possible you would be able to import your own packages into the user-script like you can for message converters.

I don't understand the last question about schema names as selectors.
